source/common/network/connection_impl.cc

```diff
@@ -434,12 +435,17 @@ void ConnectionImpl::write(Buffer::Instance& data, bool end_stream, bool through
     // might need to change if we ever copy here.
     write_buffer_->move(data);
 
-    // Activating a write event before the socket is connected has the side-effect of tricking
-    // doWriteReady into thinking the socket is connected. On macOS, the underlying write may fail
-    // with a connection error if a call to write(2) occurs before the connection is completed.
-    if (!connecting_) {
-      ASSERT(file_event_ != nullptr, "ConnectionImpl file event was unexpectedly reset");
-      file_event_->activate(Event::FileReadyType::Write);
+    // If the writing to the buffer caused an overflow-watermark breach, the connection should be
+    // closed
+    // TODO(adip): what should be done if the state is Closing
```

Comment:
I think for connection-level resets, since this is highly unusual, it should just do a hard connection shutdown.